src/semaphore.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright 2018-2024 the Deno authors. MIT license.
+
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::future::Future;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::Context;
+use std::task::Poll;
+use std::task::Waker;
+
+use crate::Flag;
+
+struct WakerWithFutureDroppedState {
+  inner: Waker,
+  // if the future is dropped then we don't want to wake
+  // this waker, but the next one in the queue
+  is_future_dropped: Rc<Flag>,
+}
+
+impl WakerWithFutureDroppedState {
+  pub fn new(inner: Waker, is_future_dropped: Rc<Flag>) -> Self {
+    Self {
+      inner,
+      is_future_dropped,
+    }
+  }
+
+  pub fn is_future_dropped(&self) -> bool {
+    self.is_future_dropped.is_raised()
+  }
+
+  pub fn wake(self) {
+    debug_assert!(
+      !self.is_future_dropped.is_raised(),
+      "Waking a future that has been dropped"
+    );
+    self.inner.wake();
+  }
+}
+
+struct SemaphoreState {
+  closed: bool,
+  max_permits: usize,
+  acquired_permits: usize,
+  wakers: VecDeque<WakerWithFutureDroppedState>,
+}
+
+pub struct SemaphorePermit(Rc<Semaphore>);
+
+impl Drop for SemaphorePermit {
+  fn drop(&mut self) {
+    self.0.release();
+  }
+}
+
+pub struct Semaphore {
+  state: RefCell<SemaphoreState>,
+}
+
+impl Semaphore {
+  pub fn new(max_permits: usize) -> Self {
+    Self {
+      state: RefCell::new(SemaphoreState {
+        closed: false,
+        max_permits,
+        acquired_permits: 0,
+        wakers: VecDeque::new(),
+      }),
+    }
+  }
+
+  pub fn acquire(
+    self: &Rc<Self>,
+  ) -> impl Future<Output = Result<SemaphorePermit, ()>> {
+    AcquireFuture {
+      semaphore: self.clone(),
+      is_future_dropped: Default::default(),
+    }
+  }
```

Comment:
An `acquire` actually doesn't take place until the future is polled. Is that ok?